File: libraries/OclComparator.swift
import Foundation
import Darwin

class OclComparator
{ private static var instance : OclComparator? = nil

  init() { }

  init(copyFrom: OclComparator) {
  }

  func copy() -> OclComparator
  { let res : OclComparator = OclComparator(copyFrom: self)
    addOclComparator(instance: res)
    return res
  }

  static func defaultInstanceOclComparator() -> OclComparator
  { if (instance == nil)
    { instance = createOclComparator() }
    return instance!
  }

  deinit
  { killOclComparator(obj: self) }


  func compare<T: Comparable>(lhs : T, rhs : T) -> Int
  {
    var result : Int = 0
    result = (((lhs < rhs) ? -1 : ((lhs > rhs) ? 1 : 0)))
    return result

  }

  static func lowerSegment<T : Comparable>(col : [T], x : T, cmp : OclComparator) -> [T]
  {
    var result : [T] = []
    result = Ocl.select(s: col, f: { y in cmp.compare(lhs: y, rhs: x) < 0 })
    return result
  }
  
  static func binarySearch<T : Comparable>(col : [T], x : T, cmp : OclComparator) -> Int
  {
    var result : Int = 0
    result = OclComparator.lowerSegment(col: col, x: x, cmp: cmp).count
    return result
  }

  static func sortWith<T : Comparable>(col : [T], cmp : OclComparator) -> [T]
  {
    var result : [T] = []
    result = Ocl.sortedBy(s: col, f: { x in  OclComparator.lowerSegment(col: col, x: x, cmp: cmp).count })
    return result
  }

  static func maximumWith<T : Comparable>(col : [T], cmp : OclComparator) -> T
  { 
    for x in col { 
      if Ocl.forAll(s: col, f: { y in cmp.compare(lhs: y, rhs: x) <= 0 })
      { return x }
    }
    return Ocl.any(s: col)!
  }


  static func minimumWith<T : Comparable>(col : [T], cmp : OclComparator) -> T
  { for x in col { 
      if Ocl.forAll(s: col, f: { y in cmp.compare(lhs: y, rhs: x) >= 0 })
      { return x }
    }
    return Ocl.any(s: col)!
  }

} 

var OclComparator_allInstances : [OclComparator] = [OclComparator]()

func createOclComparator() -> OclComparator
{ let result : OclComparator = OclComparator()
  OclComparator_allInstances.append(result)
  return result
}

func addOclComparator(instance : OclComparator)
{ OclComparator_allInstances.append(instance) }

func killOclComparator(obj: OclComparator)
{ OclComparator_allInstances = OclComparator_allInstances.filter{ $0 !== obj } }

File: libraries/OclComparator.cpp
int OclComparator::compare(void* lhs, void* rhs)
{
    int result = 0;
    result = (lhs < rhs) ? -1 : ((rhs < lhs) ? 1 : 0);
    return result;
}


vector<void*>* OclComparator::lowerSegment(vector<void*>* col, void* x, OclComparator* cmp)
{
    vector<void*>* result = new vector<void*>();
    result = select_0(col, cmp, x);
    return result;
}

vector<void*>* OclComparator::sortWith(vector<void*>* col, OclComparator* cmp)
{
    vector<void*>* result = new vector<void*>();
    result = UmlRsdsLib<void*>::sort(col, [=](void* _var1, void* _var2) -> bool{ return OclComparator::lowerSegment(col,_var1,cmp)->size() < OclComparator::lowerSegment(col,_var2,cmp)->size(); });
    return result;
}

void* OclComparator::maximumWith(vector<void*>* col, OclComparator* cmp)
{
    void* result = NULL;
    if (col->size() <= 0) { return result; }
    result = UmlRsdsLib<void*>::any(select_2(col, cmp, col));
    return result;
}


void* OclComparator::minimumWith(vector<void*>* col, OclComparator* cmp)
{
    void* result = NULL;
    if (col->size() <= 0) { return result; }
    result = UmlRsdsLib<void*>::any(select_4(col, cmp, col));
    return result;
}


int OclComparator::binarySearch(vector<void*>* col, void* x, OclComparator* cmp)
{
    int result = 0;
    result = OclComparator::lowerSegment(col, x, cmp)->size();
    return result;
}

